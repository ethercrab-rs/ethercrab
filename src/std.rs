use crate::{client::Client, timer_factory::TimerFactory};
use core::{future::Future, task::Poll};
use futures_lite::FutureExt;
use pnet::datalink::{self, DataLinkReceiver, DataLinkSender};
use std::sync::Arc;

pub fn get_tx_rx(
    device: &str,
) -> Result<(Box<dyn DataLinkSender>, Box<dyn DataLinkReceiver>), std::io::Error> {
    let interfaces = datalink::interfaces();

    dbg!(&interfaces);

    let interface = interfaces
        .into_iter()
        .find(|interface| interface.name == device)
        .expect("Could not find interface");

    dbg!(interface.mac);

    let (tx, rx) = match datalink::channel(&interface, Default::default()) {
        Ok(datalink::Channel::Ethernet(tx, rx)) => (tx, rx),
        // FIXME
        Ok(_) => panic!("Unhandled channel type"),
        Err(e) => return Err(e),
    };

    Ok((tx, rx))
}

// TODO: Proper error - there are a couple of unwraps in here
// TODO: Make some sort of split() method to ensure we can only ever have one tx/rx future running
pub fn tx_rx_task<
    'a,
    const MAX_FRAMES: usize,
    const MAX_PDU_DATA: usize,
    const MAX_SLAVES: usize,
    TIMEOUT,
>(
    device: &str,
    client: &'a Arc<Client<MAX_FRAMES, MAX_PDU_DATA, MAX_SLAVES, TIMEOUT>>,
) -> Result<impl Future<Output = ()>, std::io::Error>
where
    TIMEOUT: TimerFactory + Send + 'static,
{
    let client_tx = client.clone();
    let client_rx = client.clone();

    let (mut tx, mut rx) = get_tx_rx(device)?;

    let mut packet_buf = [0u8; 1536];

    // TODO: Unwraps
    let tx_task = futures_lite::future::poll_fn::<(), _>(move |ctx| {
        client_tx
            .pdu_loop
            .send_frames_blocking(ctx.waker(), |pdu| {
                let packet = pdu.to_ethernet_frame(&mut packet_buf).unwrap();

                tx.send_to(packet, None).unwrap().map_err(|_| ())
            })
            .unwrap();

        Poll::Pending
    });

    // TODO: Unwraps
    let rx_task = smol::unblock(move || {
        loop {
            match rx.next() {
                Ok(packet) => {
                    client_rx
                        .pdu_loop
                        .pdu_rx(packet)
                        .map_err(|e| {
                            trace!("Packet len {}", packet.len());

                            e
                        })
                        .expect("RX");
                }
                Err(e) => {
                    // If an error occurs, we can handle it here
                    panic!("An error occurred while reading: {}", e);
                }
            }
        }
    });

    Ok(tx_task.race(rx_task))
}
