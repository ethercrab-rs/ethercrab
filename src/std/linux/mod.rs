//! Items to use when not in `no_std` environments.

mod raw_socket;

use self::raw_socket::RawSocketDesc;
use crate::{
    error::Error,
    fmt,
    pdu_loop::{PduRx, PduTx},
};
use async_io::Async;
use core::{future::Future, pin::Pin, task::Poll};
use futures_lite::io::{AsyncRead, AsyncWrite};

struct TxRxFut<'a> {
    socket: Async<RawSocketDesc>,
    tx: PduTx<'a>,
    rx: PduRx<'a>,
}

impl Future for TxRxFut<'_> {
    type Output = Result<(), Error>;

    fn poll(mut self: Pin<&mut Self>, ctx: &mut core::task::Context<'_>) -> Poll<Self::Output> {
        let mut buf = [0u8; 1536];

        // Lock the waker so we don't poll concurrently. spin::RwLock does this atomically
        let mut waker = self.tx.waker();

        while let Some(frame) = self.tx.next_sendable_frame() {
            let res = frame.send_blocking(&mut buf, |data| {
                #[cfg(feature = "bench-hacks")]
                {
                    // Epic hack to make data writable
                    let data: &mut [u8] = unsafe {
                        core::slice::from_raw_parts_mut(data.as_ptr() as *mut u8, data.len())
                    };

                    let mut frame = smoltcp::wire::EthernetFrame::new_unchecked(data);

                    // For benchmarks, change the first octet of the source MAC address so the packet
                    // isn't filtered out as a sent frame but is treated as a received frame instead.
                    frame.set_src_addr(smoltcp::wire::EthernetAddress([
                        0x12, 0x10, 0x10, 0x10, 0x10, 0x10,
                    ]))
                }

                match Pin::new(&mut self.socket).poll_write(ctx, data) {
                    Poll::Ready(Ok(bytes_written)) => {
                        if bytes_written != data.len() {
                            fmt::error!("Only wrote {} of {} bytes", bytes_written, data.len());

                            Err(Error::PartialSend {
                                len: data.len(),
                                sent: bytes_written,
                            })
                        } else {
                            Ok(bytes_written)
                        }
                    }

                    Poll::Ready(Err(e)) => {
                        fmt::error!("Send PDU failed: {}", e);

                        Err(Error::SendFrame)
                    }
                    Poll::Pending => Ok(0),
                }
            });

            if let Err(e) = res {
                fmt::error!("Send PDU failed: {}", e);

                return Poll::Ready(Err(e));
            }
        }

        match Pin::new(&mut self.socket).poll_read(ctx, &mut buf) {
            Poll::Ready(Ok(n)) => {
                let packet = &buf[0..n];

                if n == 0 {
                    fmt::warn!("Received zero bytes");
                }

                if let Err(e) = self.rx.receive_frame(packet) {
                    fmt::error!("Failed to receive frame: {}", e);

                    return Poll::Ready(Err(Error::ReceiveFrame));
                }
            }
            Poll::Ready(Err(e)) => {
                fmt::error!("Receive PDU failed: {}", e);

                return Poll::Ready(Err(Error::ReceiveFrame));
            }
            Poll::Pending => (),
        }

        waker.replace(ctx.waker().clone());

        Poll::Pending
    }
}

/// Spawn a TX and RX task.
pub fn tx_rx_task(
    interface: &str,
    pdu_tx: PduTx<'static>,
    pdu_rx: PduRx<'static>,
) -> Result<impl Future<Output = Result<(), Error>>, std::io::Error> {
    let socket = RawSocketDesc::new(interface)?;

    let async_socket = async_io::Async::new(socket)?;

    let task = TxRxFut {
        socket: async_socket,
        tx: pdu_tx,
        rx: pdu_rx,
    };

    Ok(task)
}
