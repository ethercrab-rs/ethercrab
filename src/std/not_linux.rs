//! Items to use when not in `no_std` environments.

use crate::{
    error::Error,
    pdu_loop::{PduRx, PduTx},
};
use core::future::Future;
use embassy_futures::select;
use pnet_datalink::{self, channel, Channel, DataLinkReceiver, DataLinkSender};
use smoltcp::wire::EthernetFrame;

/// Get a TX/RX pair.
fn get_tx_rx(
    device: &str,
) -> Result<(Box<dyn DataLinkSender>, Box<dyn DataLinkReceiver>), std::io::Error> {
    let interfaces = pnet_datalink::interfaces();

    let interface = match interfaces.iter().find(|interface| interface.name == device) {
        Some(interface) => interface,
        None => {
            log::error!("Could not find interface {device}");

            log::error!("Available interfaces:");

            for interface in interfaces.iter() {
                log::error!("-> {} {}", interface.name, interface.description);
            }

            panic!();
        }
    };

    let config = pnet_datalink::Config {
        write_buffer_size: 16384,
        read_buffer_size: 16384,
        ..Default::default()
    };

    let (tx, rx) = match channel(interface, config) {
        Ok(Channel::Ethernet(tx, rx)) => (tx, rx),
        Ok(_) => panic!("Unhandled channel type"),
        Err(e) => return Err(e),
    };

    Ok((tx, rx))
}

// TODO: Proper error - there are a couple of unwraps in here
/// Create a task that waits for PDUs to send, and receives PDU responses.
pub fn tx_rx_task(
    device: &str,
    mut pdu_tx: PduTx<'static>,
    mut pdu_rx: PduRx<'static>,
) -> Result<impl Future<Output = Result<(), Error>>, std::io::Error> {
    let (mut tx, mut rx) = get_tx_rx(device)?;

    let mut packet_buf = [0u8; 1536];

    let task = async move {
        // TODO: Unwraps
        let tx_task = async {
            loop {
                while let Some(frame) = pdu_tx.next_sendable_frame() {
                    frame
                        .send(&mut packet_buf, |frame_bytes| async {
                            tx.send_to(frame_bytes, None)
                                .unwrap()
                                .map_err(|e| {
                                    log::error!("Failed to send packet: {e}");
                                })
                                .expect("TX");

                            Ok(frame_bytes.len())
                        })
                        .await
                        .expect("TX");
                }

                futures_lite::future::yield_now().await;
            }
        };

        // TODO: Unwraps
        let rx_task = blocking::unblock(move || {
            let mut frame_buf: Vec<u8> = Vec::new();

            loop {
                match rx.next() {
                    Ok(ethernet_frame) => {
                        match EthernetFrame::new_unchecked(ethernet_frame).check_len() {
                            // We got a full frame
                            Ok(_) => {
                                if !frame_buf.is_empty() {
                                    log::warn!("{} existing frame bytes", frame_buf.len());
                                }

                                frame_buf.extend_from_slice(ethernet_frame);
                            }
                            // Truncated frame - try adding them together
                            Err(_) => {
                                log::warn!("Truncated frame: len {}", ethernet_frame.len());

                                frame_buf.extend_from_slice(ethernet_frame);

                                continue;
                            }
                        };

                        pdu_rx
                            .receive_frame(&frame_buf)
                            .map_err(|e| {
                                dbg!(frame_buf.len());

                                e
                            })
                            .expect("RX");

                        frame_buf.truncate(0);
                    }
                    Err(e) => {
                        // If an error occurs, we can handle it here
                        panic!("An error occurred while reading: {e}");
                    }
                }

                std::thread::yield_now();
            }
        });

        select::select(tx_task, rx_task).await;

        Ok(())
    };

    Ok(task)
}
