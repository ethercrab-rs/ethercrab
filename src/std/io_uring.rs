use crate::{error::Error, fmt, std::unix::RawSocketDesc, std::ParkSignal, PduRx, PduTx};
use core::{mem::MaybeUninit, task::Waker};
use io_uring::{opcode, IoUring};
use smallvec::{smallvec, SmallVec};
use std::{io, os::fd::AsRawFd, sync::Arc, time::Instant};

/// Use the upper bit of a u64 to mark whether a frame is a write (`1`) or a read (`0`).
const WRITE_MASK: u64 = 1 << 63;
const ENTRIES: usize = 256;

/// Create a blocking TX/RX loop using `io_uring`.
///
/// This function is only available on `linux` targets as it requires `io_uring` support. Older
/// kernels may not support `io_uring`.
pub fn tx_rx_task_io_uring<'sto>(
    interface: &str,
    mut pdu_tx: PduTx<'sto>,
    mut pdu_rx: PduRx<'sto>,
) -> Result<(PduTx<'sto>, PduRx<'sto>), io::Error> {
    let mut socket = RawSocketDesc::new(interface)?;

    let mtu = socket.interface_mtu()?;

    fmt::debug!(
        "Opening {} with MTU {}, blocking, using io_uring",
        interface,
        mtu
    );

    // MTU is payload size. We need to add the layer 2 header which is 18 bytes.
    let mtu = mtu + 18;

    // SAFETY: Max entries is 256 because `PduStorage::N` is checked to be in 0..u8::MAX, and will
    // eventually be a `u8` once const generics get there. Twice as much space is reserved as each
    // frame requires a send _and_ receive buffer.
    //
    // This data MUST NOT MOVE or be reordered once created as io_uring holds pointers into it.
    let mut bufs: slab::Slab<(io_uring::squeue::Entry, SmallVec<[u8; 1518]>)> =
        slab::Slab::with_capacity(ENTRIES * 2);

    let mut ring = IoUring::new(ENTRIES as u32)?;

    let mut high_water_mark = 0;

    let signal = Arc::new(ParkSignal::new());
    let waker = Waker::from(Arc::clone(&signal));

    loop {
        pdu_tx.replace_waker(&waker);

        let mut sent = 0;

        while let Some(frame) = pdu_tx.next_sendable_frame() {
            let idx = frame.storage_slot_index();

            let tx_b = bufs.vacant_entry();
            let tx_key = tx_b.key();
            let (tx_entry, tx_buf) = tx_b.insert((
                unsafe { MaybeUninit::zeroed().assume_init() },
                smallvec![0; mtu],
            ));

            frame
                .send_blocking(|data: &[u8]| {
                    *tx_entry = opcode::Write::new(
                        io_uring::types::Fd(socket.as_raw_fd()),
                        data.as_ptr(),
                        data.len() as _,
                    )
                    .build()
                    // Distinguish sent frames from received frames by using the upper bit of
                    // the user data as a flag.
                    .user_data(tx_key as u64 | WRITE_MASK);

                    // TODO: Zero copy
                    tx_buf
                        .get_mut(0..data.len())
                        .ok_or(Error::Internal)?
                        .copy_from_slice(data);

                    while unsafe { ring.submission().push(tx_entry).is_err() } {
                        // If the submission queue is full, flush it to the kernel
                        ring.submit().expect("Internal error, failed to submit ops");
                    }

                    sent += 1;

                    Ok(data.len())
                })
                .expect("Send blocking");

            let rx_b = bufs.vacant_entry();
            let rx_key = rx_b.key();
            let (rx_entry, rx_buf) = rx_b.insert((
                unsafe { MaybeUninit::zeroed().assume_init() },
                smallvec![0; mtu],
            ));

            *rx_entry = opcode::Read::new(
                io_uring::types::Fd(socket.as_raw_fd()),
                rx_buf.as_mut_ptr() as _,
                rx_buf.len() as _,
            )
            .build()
            .user_data(rx_key as u64);

            fmt::trace!(
                "Insert frame TX {:#04x}, key {}, RX key {}",
                idx,
                tx_key,
                rx_key
            );

            while unsafe { ring.submission().push(rx_entry).is_err() } {
                // If the submission queue is full, flush it to the kernel
                ring.submit().expect("Internal error, failed to submit ops");
            }

            high_water_mark = high_water_mark.max(bufs.len());
        }

        // TODO: Collect these metrics for later gathering instead of just asserting
        // assert_eq!(ring.completion().overflow(), 0);
        // assert_eq!(ring.completion().is_full(), false);
        // assert_eq!(ring.submission().cq_overflow(), false);
        // assert_eq!(ring.submission().dropped(), 0);

        ring.submission().sync();

        let now = Instant::now();

        if sent > 0 {
            ring.submit_and_wait(sent * 2)?;
        }

        fmt::trace!(
            "Submitted, waited for {} completions for {} us",
            ring.completion().len(),
            now.elapsed().as_micros(),
        );

        // SAFETY: We must never call `completion_shared` or `completion` inside this loop.
        for recv in unsafe { ring.completion_shared() } {
            if recv.result() < 0 && recv.result() != -libc::EWOULDBLOCK {
                return Err(io::Error::last_os_error());
            }

            let key = recv.user_data();

            let received = Instant::now();

            fmt::trace!(
                "Got a frame by key {} -> {} {}",
                key,
                key & !WRITE_MASK,
                if key & WRITE_MASK == WRITE_MASK {
                    "---->"
                } else {
                    "<--"
                }
            );

            // If upper bit is set, this was a write that is now complete. We can remove its buffer
            // from the slab allocator.
            if key & WRITE_MASK == WRITE_MASK {
                let key = key & !WRITE_MASK;

                // Clear send buffer grant as it's been sent over the network
                bufs.remove(key as usize);

                continue;
            }

            // Original read did not succeed. Requeue read so we can try again.
            if recv.result() == -libc::EWOULDBLOCK {
                fmt::trace!("Frame key {} would block. Queuing for retry", key);

                let (rx_entry, _buf) = bufs.get(key as usize).expect("Could not get retry entry");

                // SAFETY: `submission_shared` must not be held at the same time this one is
                while unsafe { ring.submission_shared().push(rx_entry).is_err() } {
                    // If the submission queue is full, flush it to the kernel
                    ring.submit().expect("Internal error, failed to submit ops");
                }
            } else {
                let (_entry, frame) = bufs.remove(key as usize);

                let frame_index = frame
                    .get(0x11)
                    .ok_or_else(|| io::Error::other(Error::Internal))?;

                fmt::trace!(
                    "Raw frame {:#04x} result {} buffer key {}",
                    frame_index,
                    recv.result(),
                    key,
                );

                pdu_rx.receive_frame(&frame).map_err(io::Error::other)?;

                fmt::trace!("Received frame in {} ns", received.elapsed().as_nanos());
            }
        }

        if bufs.is_empty() {
            fmt::trace!("No frames in flight, waiting to be woken with new frames to send");

            let start = Instant::now();

            // This must be after the send packet code as there can be a (safe!) race condition on
            // startup where the TX waker hasn't been registered yet, so when a future from another
            // thread tries to send its frame, it has no waker, so we just end up waiting forever.
            //
            // If this wait() is down here, we get at least one loop where any queued packets can be
            // sent.
            signal.wait();

            fmt::trace!("--> Waited for {} ns", start.elapsed().as_nanos());

            if pdu_tx.should_exit() {
                fmt::debug!("io_uring TX/RX was asked to exit");

                return Ok((pdu_tx.release(), pdu_rx.release()));
            }
        } else {
            fmt::trace!(
                "Buf keys {:?} in flight",
                bufs.iter().map(|(k, _v)| k).collect::<Vec<_>>(),
            );
        }
    }
}
