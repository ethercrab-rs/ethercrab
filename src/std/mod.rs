//! Items required for running in `std` environments.

#[cfg(target_os = "linux")]
mod io_uring;
#[cfg(unix)]
mod unix;
#[cfg(target_os = "windows")]
mod windows;
#[cfg(all(target_os = "linux", feature = "xdp"))]
mod xdp;

use std::{
    sync::Arc,
    task::Wake,
    thread::{self, Thread},
};

#[cfg(target_os = "windows")]
pub use self::windows::{TxRxTaskConfig, ethercat_now, tx_rx_task_blocking};
#[cfg(unix)]
pub use unix::{ethercat_now, tx_rx_task};
// io_uring is Linux-only
#[cfg(target_os = "linux")]
pub use io_uring::tx_rx_task_io_uring;
#[cfg(all(target_os = "linux", feature = "xdp"))]
pub use xdp::tx_rx_task_xdp;

struct ParkSignal {
    current_thread: Thread,
}

impl ParkSignal {
    fn new() -> Self {
        Self {
            current_thread: thread::current(),
        }
    }

    fn wait(&self) {
        thread::park();
    }

    // fn wait_timeout(&self, timeout: Duration) {
    //     thread::park_timeout(timeout)
    // }
}

impl Wake for ParkSignal {
    fn wake(self: Arc<Self>) {
        self.current_thread.unpark();
    }
}
