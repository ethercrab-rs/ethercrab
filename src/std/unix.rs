//! Items to use when not in `no_std` environments.

use crate::{
    error::Error,
    pdu_loop::{PduRx, PduTx},
};
use core::future::Future;
use embassy_futures::select;
use smoltcp::phy::{Device, Medium, RxToken, TxToken};
use std::io;

/// Spawn a TX and RX task.
pub fn tx_rx_task(
    interface: &str,
    mut pdu_tx: PduTx<'static>,
    mut pdu_rx: PduRx<'static>,
) -> Result<impl Future<Output = Result<(), Error>>, std::io::Error> {
    let socket = smoltcp::phy::RawSocket::new(interface, Medium::Ethernet)?;
    let socket_rx = smoltcp::phy::RawSocket::new(interface, Medium::Ethernet)?;

    let mut async_socket = smol::Async::new(socket)?;
    let mut async_socket_rx = smol::Async::new(socket_rx)?;

    let mut packet_buf = [0u8; 1536];

    let task = async move {
        let tx_task = async {
            loop {
                let frames = pdu_tx.next().await;

                for frame in frames {
                    frame
                        .send(&mut packet_buf, |frame_bytes| async {
                            async_socket
                                .write_with_mut(|iface| {
                                    // transmit() always returns Some() (at time of writing), so this unwrap
                                    // should optimise out.
                                    let tx_token = iface.transmit().unwrap();

                                    tx_token
                                        .consume(
                                            smoltcp::time::Instant::now(),
                                            frame_bytes.len(),
                                            |buf| {
                                                buf.copy_from_slice(frame_bytes);

                                                Ok(())
                                            },
                                        )
                                        .expect("consume");

                                    Ok(())
                                })
                                .await
                                .map_err(|e| {
                                    log::error!("Send failed: {}", e);

                                    Error::SendFrame
                                })
                        })
                        .await
                        .expect("TX");
                }
            }
        };

        let rx_task = async {
            async_socket_rx
                .read_with_mut(|iface| {
                    if let Some((receiver, _)) = iface.receive() {
                        receiver
                            .consume(smoltcp::time::Instant::now(), |frame_buf| {
                                pdu_rx.receive_frame(&frame_buf).map_err(|e| {
                                    log::error!("Failed to receive packet: {}", e);

                                    smoltcp::Error::Malformed
                                })
                            })
                            .expect("RX packet");
                    }

                    Err::<(), _>(io::Error::from(io::ErrorKind::WouldBlock))
                })
                .await
                .expect("Read loop");
        };

        select::select(tx_task, rx_task).await;

        Ok(())
    };

    Ok(task)
}
