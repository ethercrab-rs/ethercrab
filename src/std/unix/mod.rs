//! Items to use when not in `no_std` environments.

#[cfg(all(not(target_os = "linux"), unix))]
mod bpf;
#[cfg(target_os = "linux")]
mod linux;

#[cfg(all(not(target_os = "linux"), unix))]
use self::bpf::BpfDevice as RawSocketDesc;
#[cfg(target_os = "linux")]
pub(in crate::std) use self::linux::RawSocketDesc;

use crate::{
    error::{Error, PduError},
    fmt,
    pdu_loop::{PduRx, PduTx},
};
use async_io::Async;
use core::{future::Future, pin::Pin, task::Poll};
use futures_lite::{AsyncRead, AsyncWrite};
use std::thread;

struct TxRxFut<'a> {
    socket: Async<RawSocketDesc>,
    mtu: usize,
    tx: PduTx<'a>,
    rx: PduRx<'a>,
}

impl Future for TxRxFut<'_> {
    type Output = Result<(), Error>;

    fn poll(mut self: Pin<&mut Self>, ctx: &mut core::task::Context<'_>) -> Poll<Self::Output> {
        // Re-register waker to make sure this future is polled again
        self.tx.replace_waker(ctx.waker());

        while let Some(frame) = self.tx.next_sendable_frame() {
            let res = frame.send_blocking(|data| {
                match Pin::new(&mut self.socket).poll_write(ctx, data) {
                    Poll::Ready(Ok(bytes_written)) => {
                        if bytes_written != data.len() {
                            fmt::error!("Only wrote {} of {} bytes", bytes_written, data.len());

                            Err(Error::PartialSend {
                                len: data.len(),
                                sent: bytes_written,
                            })
                        } else {
                            Ok(bytes_written)
                        }
                    }

                    Poll::Ready(Err(e)) => {
                        fmt::error!("Send PDU failed: {}", e);

                        Err(Error::SendFrame)
                    }
                    Poll::Pending => Ok(0),
                }
            });

            if let Err(e) = res {
                fmt::error!("Send PDU failed: {}", e);

                return Poll::Ready(Err(e));
            }
        }

        let mut buf = vec![0; self.mtu];

        match Pin::new(&mut self.socket).poll_read(ctx, &mut buf) {
            Poll::Ready(Ok(n)) => {
                fmt::trace!("Poll ready");
                // Wake again in case there are more frames to consume. This is additionally
                // important for macOS as multiple packets may be received for one `poll_read`
                // call, but will only be returned during the _next_ `poll_read`. If this line
                // is removed, PDU response frames are missed, causing timeout errors.
                ctx.waker().wake_by_ref();

                let packet = buf.get(0..n).ok_or(Error::Internal)?;

                if n == 0 {
                    fmt::warn!("Received zero bytes");
                }

                loop {
                    match self.rx.receive_frame(packet) {
                        // Wait for frame RX future waker to be registered
                        Err(Error::Pdu(PduError::NoWaker)) => thread::yield_now(),
                        Err(e) => {
                            fmt::error!("Failed to receive frame: {}", e);

                            return Poll::Ready(Err(Error::ReceiveFrame));
                        }
                        Ok(_) => break,
                    }
                }
            }
            Poll::Ready(Err(e)) => {
                fmt::error!("Receive PDU failed: {}", e);
            }
            Poll::Pending => (),
        }

        Poll::Pending
    }
}

/// Spawn a TX and RX task.
pub fn tx_rx_task<'sto>(
    interface: &str,
    pdu_tx: PduTx<'sto>,
    #[allow(unused_mut)] mut pdu_rx: PduRx<'sto>,
) -> Result<impl Future<Output = Result<(), Error>> + 'sto, std::io::Error> {
    let mut socket = RawSocketDesc::new(interface)?;

    // macOS forcibly sets the source address to the NIC's MAC, so instead of using `MASTER_ADDR`
    // for filtering returned packets, we must set the address to compare to the NIC MAC.
    #[cfg(all(not(target_os = "linux"), unix))]
    if let Some(mac) = socket.mac().ok().flatten() {
        fmt::debug!("Setting source MAC to {}", mac);

        pdu_rx.set_source_mac(mac);
    }

    let mtu = socket.interface_mtu()?;

    fmt::debug!("Opening {} with MTU {}", interface, mtu);

    let async_socket = Async::new(socket)?;

    let task = TxRxFut {
        socket: async_socket,
        mtu,
        tx: pdu_tx,
        rx: pdu_rx,
    };

    Ok(task)
}

/// Get the current time in nanoseconds from the EtherCAT epoch, 2000-01-01.
///
/// On POSIX systems, this function uses the monotonic clock provided by the system.
pub fn ethercat_now() -> u64 {
    let mut time = libc::timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };

    unsafe {
        libc::clock_gettime(libc::CLOCK_MONOTONIC, &mut time);
    };

    let t = (time.tv_sec as u64) * 1_000_000_000 + (time.tv_nsec as u64);

    // EtherCAT epoch is 2000-01-01
    t.saturating_sub(946684800)
}

// Unix only
#[allow(trivial_numeric_casts)]
fn ifreq_for(name: &str) -> ifreq {
    let mut ifreq = ifreq {
        ifr_name: [0; libc::IF_NAMESIZE],
        ifr_data: 0,
    };
    for (i, byte) in name.as_bytes().iter().enumerate() {
        ifreq.ifr_name[i] = *byte as libc::c_char;
    }
    ifreq
}

#[repr(C)]
#[derive(Debug)]
#[allow(non_camel_case_types)]
struct ifreq {
    ifr_name: [libc::c_char; libc::IF_NAMESIZE],
    ifr_data: libc::c_int, /* ifr_ifindex or ifr_mtu */
}
