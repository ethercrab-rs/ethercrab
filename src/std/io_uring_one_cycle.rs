use crate::{PduRx, PduTx, error::Error, fmt, std::ParkSignal, std::unix::RawSocketDesc};
use core::{mem::MaybeUninit, task::Waker};
use io_uring::{IoUring, opcode};
use smallvec::{SmallVec, smallvec};
use std::{io, os::fd::AsRawFd, sync::Arc, time::Instant};

/// Use the upper bit of a u64 to mark whether a frame is a write (`1`) or a read (`0`).
const WRITE_MASK: u64 = 1 << 63;
const ENTRIES: usize = 256;

pub struct TX_RX_TASK_IO_URING {
    socket_desc: RawSocketDesc,
    mtu: usize,
    signal: Arc<ParkSignal>,
    bufs: slab::Slab<(io_uring::squeue::Entry, SmallVec<[u8; 1518]>)>,
    ring: IoUring,
    waker: Waker,
}

pub fn setup_tx_rx_task(interface: &str) -> Result<TX_RX_TASK_IO_URING, io::Error> {
    let mut socket = RawSocketDesc::new(interface)?;
    // MTU is payload size. We need to add the layer 2 header which is 18 bytes.
    let mtu = socket.interface_mtu()?;
    let mtu = mtu + 18;

    let mut bufs: slab::Slab<(io_uring::squeue::Entry, SmallVec<[u8; 1518]>)> =
        slab::Slab::with_capacity(ENTRIES * 2);

    let mut ring = IoUring::new(ENTRIES as u32)?;

    // checks io_uring support for used opcodes
    let mut probe = io_uring::register::Probe::new();
    ring.submitter().register_probe(&mut probe)?;
    if !(probe.is_supported(opcode::Read::CODE) && probe.is_supported(opcode::Write::CODE)) {
        log::error!("io_uring does not support read and/or write opcodes");
        return Err(io::Error::other(Error::Internal));
    }

    let mut high_water_mark = 0;

    let signal = Arc::new(ParkSignal::new());
    let waker = Waker::from(Arc::clone(&signal));

    return Ok(TX_RX_TASK_IO_URING {
        socket_desc: socket,
        mtu,
        signal: signal,
        bufs,
        ring,
        waker,
    });
}

/// Create a blocking TX/RX loop using `io_uring`.
///
/// This function is only available on `linux` targets as it requires `io_uring` support. Older
/// kernels may not support `io_uring`.
pub fn tx_rx_task_io_uring_cycle<'sto>(
    config: &mut TX_RX_TASK_IO_URING,
    mut pdu_tx: PduTx<'sto>,
    mut pdu_rx: PduRx<'sto>,
) -> Result<(PduTx<'sto>, PduRx<'sto>), io::Error> {
    pdu_tx.replace_waker(&config.waker);

    let mut sent = 0;

    // === submit all TX frames and corresponding RX frames ===
    while let Some(frame) = pdu_tx.next_sendable_frame() {
        let tx_b = config.bufs.vacant_entry();
        let tx_key = tx_b.key();
        let (tx_entry, tx_buf) = tx_b.insert((
            unsafe { MaybeUninit::zeroed().assume_init() },
            smallvec![0; config.mtu],
        ));

        let res = frame.send_blocking(|data: &[u8]| {
            *tx_entry = opcode::Write::new(
                io_uring::types::Fd(config.socket_desc.as_raw_fd()),
                data.as_ptr(),
                data.len() as _,
            )
            .build()
            .user_data(tx_key as u64 | WRITE_MASK);

            tx_buf[..data.len()].copy_from_slice(data);

            while unsafe { config.ring.submission().push(tx_entry).is_err() } {
                config.ring.submit().expect("submit failed");
            }

            sent += 1;
            Ok(data.len())
        });

        let rx_b = config.bufs.vacant_entry();
        let rx_key = rx_b.key();
        let (rx_entry, rx_buf) = rx_b.insert((
            unsafe { MaybeUninit::zeroed().assume_init() },
            smallvec![0; config.mtu],
        ));

        *rx_entry = opcode::Read::new(
            io_uring::types::Fd(config.socket_desc.as_raw_fd()),
            rx_buf.as_mut_ptr() as _,
            rx_buf.len() as _,
        )
        .build()
        .user_data(rx_key as u64);

        while unsafe { config.ring.submission().push(rx_entry).is_err() } {
            config.ring.submit().expect("submit failed");
        }
    }

    config.ring.submission().sync();

    if sent > 0 {
        config.ring.submit_and_wait(sent * 2)?;
    }

    // === process completions once ===
    for recv in unsafe { config.ring.completion_shared() } {
        let key = recv.user_data();
        if key & WRITE_MASK == WRITE_MASK {
            config.bufs.remove((key & !WRITE_MASK) as usize);
            continue;
        }
        if recv.result() < 0 && recv.result() != -libc::EWOULDBLOCK {
            return Err(io::Error::last_os_error());
        }
        let (_entry, frame) = config.bufs.remove(key as usize);
        pdu_rx.receive_frame(&frame).map_err(io::Error::other)?;
    }

    Ok((pdu_tx, pdu_rx))
}
