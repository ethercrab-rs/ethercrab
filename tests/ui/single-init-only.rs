use ethercrab::{std::ethercat_now, MainDevice, MainDeviceConfig, PduStorage, Timeouts};
use std::thread;
use std::{sync::Arc, time::Duration};

const MAX_SUBDEVICES: usize = 16;
const MAX_PDU_DATA: usize = 1100;
const MAX_FRAMES: usize = 16;
const PDI_LEN: usize = 64;

static PDU_STORAGE: PduStorage<MAX_FRAMES, MAX_PDU_DATA> = PduStorage::new();

fn main() {
    let (_tx, _rx, pdu_loop) = PDU_STORAGE.try_split().expect("can only split once");

    let mut maindevice = Arc::new(MainDevice::new(
        pdu_loop,
        Timeouts {
            wait_loop_delay: Duration::from_millis(2),
            mailbox_response: Duration::from_millis(1000),
            ..Default::default()
        },
        MainDeviceConfig::default(),
    ));

    // We should not be able to concurrently init!
    thread::scope(|s| {
        s.spawn(|| {
            let _group = smol::block_on(
                Arc::get_mut(&mut maindevice)
                    .unwrap()
                    .init_single_group::<MAX_SUBDEVICES, PDI_LEN>(ethercat_now),
            )
            .expect("Init");
        });

        s.spawn(|| {
            let _group = smol::block_on(
                Arc::get_mut(&mut maindevice)
                    .unwrap()
                    .init_single_group::<MAX_SUBDEVICES, PDI_LEN>(ethercat_now),
            )
            .expect("Init");
        });
    });
}
