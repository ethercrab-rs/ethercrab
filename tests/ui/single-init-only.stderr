error[E0499]: cannot borrow `maindevice` as mutable more than once at a time
  --> tests/ui/single-init-only.rs:36:17
   |
26 |       thread::scope(|s| {
   |                      - has type `&'1 Scope<'1, '_>`
27 |           s.spawn(|| {
   |           -       -- first mutable borrow occurs here
   |  _________|
   | |
28 | |             let _group = smol::block_on(
29 | |                 Arc::get_mut(&mut maindevice)
   | |                                   ---------- first borrow occurs due to use of `maindevice` in closure
30 | |                     .unwrap()
...  |
33 | |             .expect("Init");
34 | |         });
   | |__________- argument requires that `maindevice` is borrowed for `'1`
35 |
36 |           s.spawn(|| {
   |                   ^^ second mutable borrow occurs here
37 |               let _group = smol::block_on(
38 |                   Arc::get_mut(&mut maindevice)
   |                                     ---------- second borrow occurs due to use of `maindevice` in closure
